src/primitives/transaction.h

```diff
@@ -351,11 +366,39 @@ class CTransaction
 
     template <typename Stream, typename Operation>
     inline void SerializationOp(Stream& s, Operation ser_action, int nType, int nVersion) {
-        READWRITE(*const_cast<int32_t*>(&this->nVersion));
+        if (ser_action.ForRead()) {
+            // When deserializing, unpack the 4 byte header to extract fOverwintered and nVersion.
+            uint32_t header;
+            READWRITE(header);
+            *const_cast<bool*>(&fOverwintered) = header >> 31;
+            *const_cast<int32_t*>(&this->nVersion) = header & 0x7FFFFFFF;
+        } else {
+            // When serializing v1 and v2, the 4 byte header is nVersion
+            uint32_t header = this->nVersion;
+            // When serializing Overwintered tx, the 4 byte header is the combination of fOverwintered and nVersion
+            if (fOverwintered) {
+                header |= 1 << 31;
```

Comment:
Does this ever change `header`? Seems you're setting the 31st bit only when it was set already.

Reply:
`this->nVersion` (the internal representation of the version number) will now always positive, and thus will never have the 31st bit set.

Reply:
OK. What I meant was that for overwinter the value of header in 372 is like the final value,
but I missed that setting it this way gives the correct value also for pre-overwinter